### PR TITLE
output unaligned FASTQ files TheiaCov_Illumina PE and SE

### LINKED ISSUE #242: [feature request] TheiaCov: output unaligned read in FASTQ & BAM format
Started dev branch `cjk-unaligned` for this. Still needs work.

Will fill out this Issue later

### PR DESCRIPTION
**Pull request by kapsakcj**
Closes #242 

This dev branch should NOT be deleted after merging, we need to notify Alex E. and recommend switching to `main` branch for further usage before deleting it

### changes to `tasks/alignment/task_bwa.wdl`

- expose `memory` and `docker` as optional inputs
- added functionality to cope with uncompressed FASTQ files used as input
- added lots of `echo` statements throughout to help with debugging and troubleshooting (we should keep these, they will be useful in the future)
- **IMPORTANT CHANGES** 
  - lines 41-48 Altered the main `bwa` command to produce a sorted SAM file as output. Also pass in `cpus` to speed things up
  - lines 51-56 added `samtools view` command to produce a BAM that ONLY includes **aligned reads**
  - lines 59-64 added `samtools view` command to produce a BAM that ONLY includes **unaligned reads**
  - lines 74-79 Altered `samtools fastq` command to produce paired FASTQ files for **aligned reads** 
  - lines 82-87 Added `samtools fastq` command to produced paired FASTQ files for **unaligned reads**
  - lines 89-94 altered `samtools fastq` command to produce FASTQ file for **aligned reads** (single end)
  - lines 95-101 added `samtools fastq` command to produce FASTQ file for **unaligned reads** (single end)
  - lines 105-106 added commands for indexing aligned BAM and unaligned BAM
- re-enabled memory retry feature (it was commented out, likely by accident)

### Changes to `workflows/theiacov/wf_theiacov_illumina_pe.wdl` and `workflows/theiacov/wf_theiacov_illumina_se.wdl` and `workflows/utilities/wf_ivar_consensus.wdl`

- added 4 (PE) or 3 (SE) new optional outputs to both theiacov_illumina workflows:
```WDL
    File? read1_unaligned = ivar_consensus.read1_unaligned
    File? read2_unaligned = ivar_consensus.read2_unaligned
    File? sorted_bam_unaligned = ivar_consensus.sorted_bam_unaligned
    File? sorted_bam_unaligned_bai = ivar_consensus.sorted_bam_unaligned_bai
```

### Impacted Workflows/Tasks

- TheiaCov_Illumina_PE_PHB (iVar subworkflow, NOT Flu track)
- TheiaCov_Illumina_SE_PHB (iVar subworkflow, NOT Flu track)
- Freyja_FASTQ (`bwa` tasked changed, but no workflow changes) - Would be good to compare the results of the same sample analyzed via v1.3.0 vs this dev branch

## :brain: Context and Rationale

A user requested that unaligned FASTQ files (i.e. the reads that do not align to the target reference genome, example: Mpox) are output from the TheiaCov_Illumina workflows. The goal is to take these reads that did not align to the reference and perform downstream analysis on them (like Kraken for taxonomic assignment, or other purposes like attempting to assemble a genome out of those reads not aligned)

In the process of addressing this request, I also added the `unaligned_bam` and it's index which is a BAM file that only contains reads that did not align to the reference (see lines 59-64 in BWA WDL task to see where this BAM is created). Not sure how this would be used, but it doesn't hurt to output this as well.
### Inputs

added 2 new optional inputs to bwa task: `memory` and `docker`

### Outputs

- added 4 new optional outputs to both theiacov_illumina workflows (served up via the ivar subworkflow):
```WDL
    File? read1_unaligned = ivar_consensus.read1_unaligned
    File? read2_unaligned = ivar_consensus.read2_unaligned
    File? sorted_bam_unaligned = ivar_consensus.sorted_bam_unaligned
    File? sorted_bam_unaligned_bai = ivar_consensus.sorted_bam_unaligned_bai
```

### Impacted Outputs

Pre-existing outputs that may be impacted (this excludes the new ones)

- `sorted_bam` which is used heavily downstream in the workflow
- `read1_aligned` & `read2_aligned`. Both existed previously, but command syntax has changed slightly so good to double check these.

### Data
Test data used were as follows
1. Simulated SARS-CoV-2 data. This was expected to map 100% to the reference SARS-CoV-2 genome, i.e. 0 unaligned reads.
2. Human data. This was expected not to align, but some reads aligned to the the reference SARS-CoV-2 genome.
3. A spiked dataset from 1. and 2. above. SARS-CoV-2 and some human reads expected to map.
4. Unaligned reads from 2. This was to test cases when nothing aligns to a reference, in which case unaligned reads file will be created but empty.

### Locally

Worked as expected
### Terra

https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/Global_tree_testing/job_history/43118990-5338-4d92-b429-19158452cf78

## :microscope: Quality checks

Pull Request (PR) checklist:
- [X] Include a description of what is in this pull request in this message.
- [X] The workflow/task has been tested locally and on Terra
- [X] The CI/CD has been adjusted and tests are passing
- [X] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

**Comment by kapsakcj** [comment]
NOTE: this change will impact Freyja_FASTQ workflow as it used the same `bwa` task

File: tasks/alignment/task_bwa.wdl
Comment (cimendes) [inline tasks/alignment/task_bwa.wdl:55]: shouldn't this file be renamed to `~{samplename}.sorted.aligned-reads.bam`?

File: tasks/alignment/task_bwa.wdl
Comment (cimendes) [inline tasks/alignment/task_bwa.wdl:76]: given that the bam files were already filtered by `samtools view`, do we need the `-F 4` flag here too? All reads in the bam file should be aligned

File: tasks/alignment/task_bwa.wdl
Comment (cimendes) [inline tasks/alignment/task_bwa.wdl:84]: Same comment as above

File: tasks/alignment/task_bwa.wdl
Comment (cimendes) [inline tasks/alignment/task_bwa.wdl:132]: I'm having an existential crisis on if we should rename `sorted_bam` and `sorted_bai` as `sorted_aligned_bam` and `sorted_aligned_bai` for readability :sweat_smile: 
Additionally, I would rename `sorted_bam_unaligned` to `sorted_unaligned_bam` so the variable has the file type at the end. It's a bit cleaner and it goes with or style-guide. Same for `sorted_bam_unaligned_bai` -> `sorted_unaligned_bai`

**Comment by cimendes** [comment]
Workflows using `bwa`:
- **TheiaCoV_Illumina_SE** and **PE** through the **wf_ivar_consensus.wdl** sub-workflow
  - Tested by @kapsakcj 
- **Freyja_FASTQ**

File: tasks/alignment/task_bwa.wdl
Comment (jrotieno) [reply tasks/alignment/task_bwa.wdl:132]: Personally I prefer names that tell a story of what has been done to the file, in which case the bam file was sorted, and then unaligned reads extracted. Happy to hear what others think and this could feed onto the style-guide.

**Comment by cimendes** [comment]
tried running theiavalidate on two runs of Freyja_FASTQ but got stuck due to unrelated issues with this PR.

A manual size comparison of the outputted bam files confirmed that the two are of the same size:
image
image

I would like to get theiavalidate results to properly check file content but given that it's not cooperating, and this is an issue that is high priority, I'm not going to hold my review on it. 

I'll wait for more feedback on the style-guide related issues I mentioned above before hitting approve but functionally this addition is working as expected for all tested workflows. Nicely done @kapsakcj and @jrotieno!

File: tasks/alignment/task_bwa.wdl
Comment (jrotieno) [reply tasks/alignment/task_bwa.wdl:38]: Were useful for ensuring the task is doing what we think it is doing. @kapsakcj, maybe we remove?

File: tasks/alignment/task_bwa.wdl
Comment (kapsakcj) [reply tasks/alignment/task_bwa.wdl:55]: I kept it as the original filename. I agree it would provide a bit more clarity to someone reading the code, but hopefully the comments are enough. 

I would prefer not to make further changes unless truly necessary since we'd like to share with our partner asap

File: tasks/alignment/task_bwa.wdl
Comment (kapsakcj) [reply tasks/alignment/task_bwa.wdl:76]: it may be unnecessary to include the -F 4 flag, I'm unsure. But it runs as expected so I'm hesitant to change & test again 😬

**Review by cimendes — APPROVED** [review]

**Comment by kapsakcj** [comment]
FYI @jrotieno @cimendes I've set this PR to draft until we hear feedback from our partner. Let's not merge until they test and provide feedback.

If we don't get any feedback in the next week or 2, then I say we press forward with merging if we're satisfied with our own testing.

It would be good to incorporate this into the next PHB release

**Pull request merged**
